extension/agenthealth/handler/useragent/useragent.go

```diff
@@ -0,0 +1,193 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package useragent
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	telegraf "github.com/influxdata/telegraf/config"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter"
+	"go.opentelemetry.io/collector/otelcol"
+	"go.uber.org/atomic"
+	"golang.org/x/exp/maps"
+
+	"github.com/aws/amazon-cloudwatch-agent/cfg/envconfig"
+	"github.com/aws/amazon-cloudwatch-agent/internal/util/collections"
+	"github.com/aws/amazon-cloudwatch-agent/internal/version"
+	"github.com/aws/amazon-cloudwatch-agent/receiver/adapter"
+)
+
+const (
+	handlerID          = "cloudwatchagent.UserAgentHandler"
+	headerKeyUserAgent = "User-Agent"
+
+	flagRunAsUser                 = "run_as_user"
+	flagContainerInsights         = "container_insights"
+	flagPulse                     = "pulse"
+	flagEnhancedContainerInsights = "enhanced_container_insights"
+	separator                     = " "
+
+	typeInputs     = "inputs"
+	typeProcessors = "processors"
+	typeOutputs    = "outputs"
+)
+
+var (
+	singleton UserAgent
+	once      sync.Once
+)
+
+type UserAgent interface {
+	SetComponents(otelCfg *otelcol.Config, telegrafCfg *telegraf.Config)
+	SetContainerInsightsFlag()
+	Header(isUsageDataEnabled bool) string
+	Listen(listener func())
+}
+
+type userAgent struct {
+	dataLock sync.Mutex
+	id       string
+
+	listenerLock sync.Mutex
+	listeners    []func()
+	isRoot       bool
+
+	inputs     collections.Set[string]
+	processors collections.Set[string]
+	outputs    collections.Set[string]
+
+	inputsStr     atomic.String
+	processorsStr atomic.String
+	outputsStr    atomic.String
+}
+
+var _ UserAgent = (*userAgent)(nil)
+
+func (ua *userAgent) SetComponents(otelCfg *otelcol.Config, telegrafCfg *telegraf.Config) {
+	ua.dataLock.Lock()
+	defer ua.dataLock.Unlock()
```

Comment:
how likely are we going to run into concurrent reads here? We only call `SetComponents` once on init right?

Reply:
Yeah you're right. I remember thinking that and then forgetting to remove it.